docs/pages/principles.rst

```diff
@@ -0,0 +1,175 @@
+create new generators when needed.
+
+Side-Effects
+~~~~~~~~~~~~
+
+``Collection`` is a helper for making transformations to input data sets.
+Even though we can technically trigger side-effects in some operations
+through a custom ``Closure``, it's better to avoid this type of usage and instead 
+use the operations for their transformative effects (use the return values).
+
+Exception handling is one scenario where we might find ourselves wanting ``Collection``
+to behave eagerly. If we want an exception to be thrown and handled in a specific function,
+during an operation, rather than when the collection is later iterated on, we can take advantage 
+of the :ref:`squash <Squash>` operation.
+
+Testing
+~~~~~~~
+
+Working with lazy evaluation can impact how we test our code. Depending on the testing
+framework used, we have a few options at our disposal when it comes to comparing collections
+objects returned by a function.
+
+``Collection`` already provides two operations which can be used for comparison:
+
+* :ref:`Equals <Equals>` - allows usage of the `assertObjectEquals`_ assertion in *PHPUnit*
+* :ref:`Same <Same>` - allows customizing precisely how elements are compared using a callback 
+
+Note that these operations will traverse both collections as part of the comparison. As such,
+any side effects triggered in our source code will be triggered during this comparison. When
+using ``equals`` in particular, we might find it useful to apply ``squash`` to the resulting
+collection object before the comparison if our test needs to assert how many times 
+a side effect is performed. 
+
+In addition to these, in *PHPSpec* we can use the `iterateAs`_ matcher to assert how our
+final collection object will iterate.
+
+The last option is to transform the collection object into an array with the :ref:`all <All>` operation
+and use any assertion that we would normally use for arrays. However, beware of the potential
+loss of data that can happen when doing this - see the operation's documentation.
+
+Usability
+---------
+
+``Collection`` and the ``Operations`` are designed with usability and flexibility in mind.
+A few key elements that contribute to this are the usage of *variadic parameters*, *custom callbacks*,
+and the fact that operations can be used both as collection object methods or *completely separately*.
+
+Variadic Parameters
+~~~~~~~~~~~~~~~~~~~
+
+Variadic parameters are used wherever possible in operations, allowing us to more succintly apply
+multiple transformations or predicates. They will *always* be evaluated by the operation as a *logical OR*.
+
+For example, the :ref:`contains <Contains>` operation allows us to easily check whether one or more
+values are contained in the collection:
+
+.. code-block:: php
+
+    $result = Collection::fromIterable(range('a', 'c'))
+        ->contains('d'); // false
+
+    $result = Collection::fromIterable(range('a', 'c'))
+        ->contains('a', 'z'); // true
```

Comment:
How about adding an example with a AND and OR using filter?

Reply:
in addition to this or instead of this example?

Reply:
In addition 👍

Reply:
done!